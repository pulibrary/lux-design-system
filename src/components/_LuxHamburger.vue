<template>
  <span class="hamburger-box">
    <span class="hamburger-inner"></span>
  </span>
</template>

<script>
/**
 * Used in conjunction with LuxMenuBar to display a menu on narrow screens.
 */
export default {
  name: "LuxHamburger",
  status: "ready",
  release: "1.0.0",
  type: "Element",
  props: {
    /**
     * The html element name used for the wrapper.
     */
    type: {
      type: String,
      default: "div",
    },
  },
}
</script>

<style lang="scss" scoped>
// Hamburger
// ==================================================
$hamburger-padding-x: 15px !default;
$hamburger-padding-y: 15px !default;
$hamburger-layer-width: 40px !default;
$hamburger-layer-height: 4px !default;
$hamburger-layer-spacing: 6px !default;
$hamburger-layer-color: #fff !default;
$hamburger-layer-border-radius: 4px !default;
$hamburger-hover-opacity: 0.7 !default;
$hamburger-active-layer-color: $hamburger-layer-color !default;
$hamburger-active-hover-opacity: $hamburger-hover-opacity !default;

$hamburger-hover-use-filter: false !default;
$hamburger-hover-filter: opacity(50%) !default;
$hamburger-active-hover-filter: $hamburger-hover-filter !default;

.hamburger {
  padding: $hamburger-padding-y $hamburger-padding-x;
  display: inline-block;
  cursor: pointer;

  transition-property: opacity, filter;
  transition-duration: 0.15s;
  transition-timing-function: linear;

  // Normalize (<button>)
  font: inherit;
  color: inherit;
  text-transform: none;
  background-color: transparent;
  border: 0;
  margin: 0;
  overflow: visible;

  &:hover {
    @if $hamburger-hover-use-filter == true {
      filter: $hamburger-hover-filter;
    } @else {
      opacity: $hamburger-hover-opacity;
    }
  }

  &.is-active {
    &:hover {
      @if $hamburger-hover-use-filter == true {
        filter: $hamburger-active-hover-filter;
      } @else {
        opacity: $hamburger-active-hover-opacity;
      }
    }

    .hamburger-inner,
    .hamburger-inner::before,
    .hamburger-inner::after {
      background-color: $hamburger-active-layer-color;
    }
  }
}

.hamburger-box {
  width: $hamburger-layer-width;
  height: $hamburger-layer-height * 3 + $hamburger-layer-spacing * 2;
  display: inline-block;
  position: relative;
}

.hamburger-inner {
  display: block;
  top: 50%;
  margin-top: calc($hamburger-layer-height / -2);

  &,
  &::before,
  &::after {
    width: $hamburger-layer-width;
    height: $hamburger-layer-height;
    background-color: $hamburger-layer-color;
    border-radius: $hamburger-layer-border-radius;
    position: absolute;
    transition-property: transform;
    transition-duration: 0.15s;
    transition-timing-function: ease;
  }

  &::before,
  &::after {
    content: "";
    display: block;
  }

  &::before {
    top: ($hamburger-layer-spacing + $hamburger-layer-height) * -1;
  }

  &::after {
    bottom: ($hamburger-layer-spacing + $hamburger-layer-height) * -1;
  }
}

.hamburger-inner {
  transition-duration: 0.075s;
  transition-timing-function: cubic-bezier(0.55, 0.055, 0.675, 0.19);

  &::before {
    transition: top 0.075s 0.12s ease, opacity 0.075s ease;
  }

  &::after {
    transition: bottom 0.075s 0.12s ease, transform 0.075s cubic-bezier(0.55, 0.055, 0.675, 0.19);
  }
}

.is-active {
  .hamburger-inner {
    transform: rotate(45deg);
    transition-delay: 0.12s;
    transition-timing-function: cubic-bezier(0.215, 0.61, 0.355, 1);

    &::before {
      top: 0;
      opacity: 0;
      transition: top 0.075s ease, opacity 0.075s 0.12s ease;
    }

    &::after {
      bottom: 0;
      transform: rotate(-90deg);
      transition: bottom 0.075s ease, transform 0.075s 0.12s cubic-bezier(0.215, 0.61, 0.355, 1);
    }
  }
}
</style>

<docs>
  ```jsx
  <div>
    <lux-hamburger></lux-hamburger>
  </div>
  ```
</docs>
