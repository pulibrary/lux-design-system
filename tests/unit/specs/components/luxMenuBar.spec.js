import { mount } from "@vue/test-utils"
import LuxMenuBar from "@/components/LuxMenuBar.vue"
import { nextTick } from "vue"
import _LuxHamburger from "@/components/_LuxHamburger.vue"

let wrapper

describe("LuxMenuBar.vue", () => {
  beforeEach(() => {
    document.body.innerHTML = `
      <div>
        <div id="app"><h1>Heading</h1></div>
      </div>
    `
    wrapper = mount(LuxMenuBar, {
      propsData: {
        active: "Foo",
        menuItems: [
          {
            name: "Foo",
            component: "Foo",
            href: "/example/",
            children: [{ name: "Baz", component: "Baz", href: "/example/" }],
          },
          {
            name: "Foo1",
            component: "Foo1",
            href: "/example/",
            children: [{ name: "Baz1", component: "Baz1", href: "/example/" }],
          },
          { name: "Bar", component: "Bar", href: "/example/" },
        ],
      },
      global: {
        components: {
          "lux-hamburger": _LuxHamburger,
        },
      },
      attachTo: document.getElementById("app"),
    })
  })

  it("should not have a 'presentation' role within the main menu", async () => {
    wrapper.setProps({ type: "main-menu" })
    await nextTick()
    expect(wrapper.element.innerHTML).not.toContain('role="presentation"')
  })

  it("should start with aria-expanded false", async () => {
    wrapper.setProps({ type: "main-menu" })
    await nextTick()
    expect(wrapper.findAll(".lux-submenu-toggle")[0].attributes("aria-expanded")).toEqual("false")
    expect(wrapper.findAll(".lux-submenu-toggle")[1].attributes("aria-expanded")).toEqual("false")
  })

  it("should be aria-expanded true after a user opens the menu for the opened menu only", async () => {
    wrapper.setProps({ type: "main-menu" })
    await nextTick()

    wrapper.find("button.lux-submenu-toggle").trigger("click")
    await nextTick()

    expect(wrapper.findAll(".lux-submenu-toggle")[0].attributes("aria-expanded")).toEqual("true")
    expect(wrapper.findAll(".lux-submenu-toggle")[1].attributes("aria-expanded")).toEqual("false")
  })

  it("should close the menu when the user presses the Esc key", async () => {
    wrapper.setProps({ type: "main-menu" })
    await nextTick()
    wrapper.find("button.lux-submenu-toggle").trigger("click")
    await nextTick()
    expect(wrapper.find(".lux-submenu-toggle").attributes("aria-expanded")).toEqual("true")

    wrapper.find("button.lux-submenu-toggle").trigger("keydown.esc")

    await nextTick()
    expect(wrapper.find(".lux-submenu-toggle").attributes("aria-expanded")).toEqual("false")
  })

  it("should be a nav element if the type prop value is 'links'", async () => {
    expect(wrapper.find("nav").exists()).toBe(true)
    expect(wrapper.find("div").exists()).toBe(false)
    wrapper.setProps({ type: "buttons" })
    await nextTick()
    expect(wrapper.find("nav").exists()).toBe(false)
    expect(wrapper.find("div").exists()).toBe(true)
  })

  it("should emit the correct event when nav-item is clicked", () => {
    const navItem = wrapper.find(".lux-nav-item")
    navItem.trigger("click")
    const emitted1 = wrapper.emitted()
    expect(Object.prototype.hasOwnProperty.call(emitted1, "menu-item-clicked")).toBe(true)
  })

  it("should emit the correct event when menu-item is clicked", async () => {
    wrapper.setProps({ type: "buttons" })
    await nextTick()
    const menuItem = wrapper.find(".lux-menu-item")
    menuItem.trigger("click")
    const emitted2 = wrapper.emitted()
    expect(Object.prototype.hasOwnProperty.call(emitted2, "menu-item-clicked")).toBe(true)
  })

  it("should emit the correct event when active menu-item changes", () => {
    wrapper.vm.localActive = "Bar"
    const e = wrapper.emitted()
    expect(Object.prototype.hasOwnProperty.call(e, "input")).toBe(true)
  })
  /* eslint-disable quotes */
  it("should properly parse menu items into a hierarchy", () => {
    const menuItemsList = [
      {
        name: "Foo",
        component: "Foo",
        href: "/example/",
        children: [{ name: "Baz", component: "Baz", href: "/example/" }],
      },
      {
        name: "Foo1",
        component: "Foo1",
        href: "/example/",
        children: [{ name: "Baz1", component: "Baz1", href: "/example/" }],
      },
      { name: "Bar", component: "Bar", href: "/example/" },
    ]
    expect(wrapper.vm.menuItems).toEqual(menuItemsList)
  })

  it("defaults to a dark theme", async () => {
    wrapper.setProps({ type: "main-menu" })
    await nextTick()
    expect(wrapper.get("nav").classes()).toContain("dark")
  })

  it("has a light theme", async () => {
    wrapper.setProps({ type: "main-menu", theme: "light" })
    await nextTick()
    expect(wrapper.get("nav").classes()).toContain("light")
  })

  it("can include arbitrary html with the unsafe_name property", async () => {
    wrapper.setProps({
      menuItems: [
        {
          name: "Foo Bar",
          unsafe_name: "<span id='my-foo'>Foo</span> <span class='my-bar'>Bar</span>",
        },
      ],
    })
    await nextTick()
    expect(wrapper.get("#my-foo").text()).toEqual("Foo")
    expect(wrapper.get(".my-bar").text()).toEqual("Bar")
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  describe("when type is main-menu", () => {
    it("emits menu-item-clicked with metadata about the clicked menu item", async () => {
      wrapper.setProps({ type: "main-menu" })
      await nextTick()
      wrapper.findAll(".lux-nav-item")[2].trigger("click")

      expect(wrapper.emitted()["menu-item-clicked"].length).toEqual(1)
      expect(wrapper.emitted()["menu-item-clicked"][0]).toEqual([
        { name: "Bar", component: "Bar", href: "/example/" },
      ])
    })
  })
})
