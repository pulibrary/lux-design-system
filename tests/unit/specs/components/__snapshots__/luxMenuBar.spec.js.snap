// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuxMenuBar.vue has the expected html structure 1`] = `
<nav
  class="lux-nav dark"
>
  <ul>
    
    <li>
      
      <a
        aria-haspopup="true"
        class="lux-has-children lux-nav-item"
        href="/example/"
        title="Foo"
      >
        
        Foo
        
      </a>
      <ul
        aria-label="submenu"
        class="lux-nav-children"
      >
        
        <li>
          <a
            class="lux-nav-item"
            href="/example/"
            title="Baz"
          >
            Baz
          </a>
        </li>
        
      </ul>
      
    </li>
    <li>
      
      <a
        aria-haspopup="true"
        class="lux-has-children lux-nav-item"
        href="/example/"
        title="Foo1"
      >
        
        Foo1
        
      </a>
      <ul
        aria-label="submenu"
        class="lux-nav-children"
      >
        
        <li>
          <a
            class="lux-nav-item"
            href="/example/"
            title="Baz1"
          >
            Baz1
          </a>
        </li>
        
      </ul>
      
    </li>
    <li>
      <a
        class="lux-nav-item"
        href="/example/"
        title="Bar"
      >
        
        Bar
        
      </a>
    </li>
    
  </ul>
</nav>
`;
